Increase reference count of state tensor in `THPGenerator_reduce` to avoid premature garbage collection in `multiprocessing` start method `"forkserver"` and `"spawn"`

ringohoffman opened the pull request:
Fixes #146828

For this script:

```python
from __future__ import annotations

import time

import torch

def worker(generator: torch.Generator):
    print(generator.get_state())

if __name__ == '__main__':
    torch.multiprocessing.set_start_method("forkserver")  # or "spawn"

    generator = torch.Generator("cpu")
    process = torch.multiprocessing.Process(target=worker, args=(generator,))
    process.start()  # process.run() does not cause a crash

    for i in range(10):
        print("Main", i)
        time.sleep(1)

    process.join()
    process.close()

```

When I add ~~`Py_INCREF(ret)`~~ `Py_INCREF(state_tensor)`, I stop getting:

```console
$ python a.py
Main 0
Main 1
Traceback (most recent call last):
  File "/home/matthew/.conda/envs/torch39/lib/python3.9/multiprocessing/forkserver.py", line 274, in main
    code = _serve_one(child_r, fds,
  File "/home/matthew/.conda/envs/torch39/lib/python3.9/multiprocessing/forkserver.py", line 313, in _serve_one
    code = spawn._main(child_r, parent_sentinel)
  File "/home/matthew/.conda/envs/torch39/lib/python3.9/multiprocessing/spawn.py", line 126, in _main
    self = reduction.pickle.load(from_parent)
  File "/home/matthew/pytorch/torch/multiprocessing/reductions.py", line 546, in rebuild_storage_fd
    storage = cls._new_shared_fd_cpu(fd, size)
RuntimeError: unable to resize file <filename not specified> to the right size: Invalid argument (22)
Main 2
Main 3
Main 4
Main 5
Main 6
Main 7
Main 8
Main 9
```

And I start getting:

```console
$ python a.py
Main 0
Main 1
tensor([  1, 209, 156,  ...,   0,   0,   0], dtype=torch.uint8)
Main 2
Main 3
Main 4
Main 5
Main 6
Main 7
Main 8
Main 9
```

cc @pbelevich

Commit:
Increase reference count of return tuple to avoid premature garbage collection of state tensor

Fixes #146828

albanD commented on torch/csrc/Generator.cpp:
```diff
@@ -259,6 +259,7 @@ static PyObject* THPGenerator_reduce(PyObject* _self, PyObject* noargs) {
   PyTuple_SET_ITEM(state.get(), 2, THPGenerator_getState(_self, nullptr));
   PyTuple_SET_ITEM(ret.get(), 2, state.release());
 
+  Py_INCREF(ret);
```
I would expect that:
- PyTupleNew returns a new ref
- THPObjectPtr steals that on construction
- The .release() below give that single ref to the returned PyObject*

?

ringohoffman replied:
The [C++ implementation of `__reduce__` for device](https://github.com/pytorch/pytorch/blob/e74fdbe6d086d0b2f0ebfddd1a3bf4d19495a830/torch/csrc/Device.cpp#L151-L177) is similar and also does not do a `Py_INCREF` on the return tuple, so I think you are right that this is not necessary. One big difference though is that this tuple contains a tensor.

2 lines up, if I change:

```python
  PyTuple_SET_ITEM(state.get(), 2, THPGenerator_getState(_self, nullptr));
```

to:

```python
  auto state_tensor = THPGenerator_getState(_self, nullptr);
  Py_INCREF(state_tensor);
  PyTuple_SET_ITEM(state.get(), 2, state_tensor);
```

It also resolves the issue. So I am fairly certain that the reference counts on this specific tensor is the source of the issue.

https://github.com/pytorch/pytorch/blob/1ae7cc41ca3c9a55faab7033e11a6115318cba8f/torch/csrc/Generator.cpp#L67-L78

We can see that the return of `THPGenerator_getState` is wrapped with `THPVariable_Wrap`, and as you told me [PyTuple_SET_ITEM](https://docs.python.org/3/c-api/tuple.html#c.PyTuple_SET_ITEM) should be stealing that reference. So is that a more proper solution--to increase the reference count on the state tensor before packing it into the tuple?

ringohoffman replied:
In the case of `set_start_method("fork")`, `__reduce__` and `__setstate__` are not actually called when spawning another process.

With the code in `main` now, without any `Py_INCREF`:

```python
import torch

generator = torch.Generator("cpu").manual_seed(2059666523504992)
torch.save(generator, "generator.pt")
```

Using `Py_REFCNT`, you can inspect the reference count of the state tensor to be 1 for the duration of `__reduce__`.

```python
import torch

generator = torch.load("generator.pt", weights_only=False)
```

The reference count of the state tensor is 2 for the duration of `__setstate__`.

---

We know that when we use `"forkeserver"` or `"spawn"` start methods with the current code, we get the error `RuntimeError: unable to resize file <filename not specified> to the right size: Invalid argument (22)`. `__setstate__` does not even run in the other process.

What we observe with this change:

```c++
  auto state_tensor = THPGenerator_getState(_self, nullptr);
  Py_INCREF(state_tensor);
  PyTuple_SET_ITEM(state.get(), 2, state_tensor);
```

Is that in `__reduce__`, the state tensor reference count increases from 1 to 2. You no longer get the error, and `__setstate__` is actually able to run in the other process. The reference count of `state_tensor` is 2 the entire time, exactly as in the case of `torch.load` without `Py_INCREF`. 

Additionally, whether or not you `Py_INCREF` the state tensor, the reference count of the `state` 3-ple in `__setstate__` is also 2.

Commit:
Only Py_INCREF the state tensor

conversation: https://github.com/pytorch/pytorch/pull/147907#discussion_r1976031723

ringohoffman commented:
@pytorchbot label "module: random"

ringohoffman commented:
@pytorchbot label:"release notes: cpp"

ringohoffman commented:
@pytorchbot label "release notes: cpp"

albanD commented:
Hoooo I know what's happening here!
The reduce returns a temporary Tensor object. But this Tensor is deleted before the process on the other side can read it. So it fails to read it...

So another way to fix this issue is to use `torch.multiprocessing.set_sharing_strategy("file_system")` which makes the send happen via a copy and so will not have this issue.

I think the only two options we have here are:
- Not passing the state as a Tensor but something else that is always copied. Converting it to a tuple of numbers maybe?
- Add a special flag on Tensor to mark it as never to be passed across process with shared memory. This sounds like the right thing to do but might be a bit more work.
